### PR TITLE
Clean and polish before release

### PR DESCRIPTION
**Pull request by alexcombessie**
[ch53489](https://app.clubhouse.io/dataiku/story/53489/initial-release)

### CODE REVIEW

**Comment by alexcombessie** [comment]
Hi @tdesfont,

This PR is ready for you review. Here is a summary of important changes:
- Plugin name change (validated by the product team) ➡️  to make the plugin more "marketable"
- Code formatting with black ➡️  to make it consistent with other plugins
- Naming (+ type hints) of variables ➡️  to make the code easier to understand & maintain
- UX improvements ➡️  to make the interface as "self-explanatory" as possible
- Data loader refactoring ➡️  to handle multiple columns as input
- Model loading/saving refactoring ➡️  to remove the reliance on local filesystem folder and allow for Kubernetes runtime
- Add distance to the output ➡️  to answer a request from Juniper

Let me know if you have any questions!

Alex

File: python-lib/utils.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import functools
+from typing import Callable, AnyStr
+from time import perf_counter
+
+
+def time_logging(log_message: AnyStr):
+    """Decorator to log timing with a custom message"""
+
+    def inner_function(function: Callable):
+        @functools.wraps(function)
```
Comment (tdesfont) [inline python-lib/utils.py:14]: Could you explain what is the design of this function ?

**Review by tdesfont — APPROVED** [review]

**Comment by tdesfont** [comment]
Need to make some tests before merge

**Comment by tdesfont** [comment]
Seems that I can not install annoy 1.17.0 on dku34

File: custom-recipes/similarity-search-index/recipe.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+params = load_indexing_recipe_params()
+
+# Load data into vector format for indexing
+columns = [params["unique_id_column"]] + params["feature_columns"]
+input_df = params["input_dataset"].get_dataframe(columns=columns, infer_with_pandas=False)
+data_loader = DataLoader(params["unique_id_column"], params["feature_columns"])
+(vector_ids, vectors) = data_loader.convert_df_to_vectors(input_df)
+
+# Build index and save index file to output folder
+nearest_neighbor = NearestNeighborSearch(num_dimensions=vectors.shape[1], **params)
```
[the remark this answers is not in the document]
Comment (alexcombessie) [reply custom-recipes/similarity-search-index/recipe.py:21]: That's already taken care of by `DataLoader. convert_df_to_vectors`

File: custom-recipes/similarity-search-index/recipe.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+with NamedTemporaryFile() as tmp:
+    nearest_neighbor.build_save_index(vectors=vectors, index_path=tmp.name)
```
[the remark this answers is not in the document]
Comment (alexcombessie) [reply custom-recipes/similarity-search-index/recipe.py:23]: Good idea, let me add that

File: custom-recipes/similarity-search-index/recipe.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    params["index_folder"].upload_stream(nearest_neighbor.INDEX_FILE_NAME, tmp)
+
+# Save vector data and indexing config to guarantee reproducibility
+save_array_to_folder(array=vector_ids, path=nearest_neighbor.VECTOR_IDS_FILE_NAME, folder=params["index_folder"])
```
[the remark this answers is not in the document]
Comment (alexcombessie) [reply custom-recipes/similarity-search-index/recipe.py:27]: Not sure what other name would explain it? The goal is literally to save a numpy array to a Dataiku folder.

File: python-lib/data_loader.py [hunk 5]
```diff
@@ -1,29 +1,75 @@
...
+    def __init__(self, unique_id_column: AnyStr, feature_columns: List[AnyStr]):
+        self.unique_id_column = unique_id_column
+        self.feature_columns = feature_columns
+
+    @staticmethod
+    def load_vector_from_string(string: AnyStr) -> np.array:
+        """Attempt to load a stringified vector into a numpy.array"""
+        parsed = json.loads(string)
+        if isinstance(parsed, list):
+            vector = np.array(parsed).astype(np.float32)
```
[the remark this answers is not in the document]
Comment (alexcombessie) [reply python-lib/data_loader.py:27]: The underlying libraries require float32:
- Annoy: https://github.com/spotify/annoy/issues/200#issuecomment-293683192
- Faiss: https://github.com/facebookresearch/faiss/issues/590#issuecomment-421070442

File: python-lib/data_loader.py [hunk 6]
```diff
@@ -1,29 +1,75 @@
...
+        """Convert a DataFrame into the vector format required by Similarity Search algorithms"""
+        start = perf_counter()
+        self._validate_df(df)
+        if verbose:
+            logging.info(
+                f"Loading dataframe of {len(df.index)} rows and "
+                + f"{len(self.feature_columns)} column(s) into vector format..."
+            )
+        vector_ids = df[self.unique_id_column].values
+        vectors = np.empty(shape=(len(df.index), self.MAX_VECTOR_LENGTH))
```
[the remark this answers is not in the document]
Comment (alexcombessie) [reply python-lib/data_loader.py:52]: I don't know in advance how many dimensions there will be. That's why I initialize an empty array with fixed size, and then assign slices of the array. In general that's a way more efficient method than appending to an array. 
Note that np.empty is specifically optimized for that as it only pre-allocates memory for the index.

File: python-lib/utils.py (hunk 1 above)
Comment (alexcombessie) [reply python-lib/utils.py:14]: It's a standard timing and logging decorator, it's used as such: https://github.com/dataiku/dss-plugin-similarity-search/blob/a81a6e18a9eebefbfd3c9b275dce70b09a68a3bf/python-lib/nearest_neighbor/annoy.py#L34

File: custom-recipes/similarity-search-index/recipe.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexcombessie) [reply custom-recipes/similarity-search-index/recipe.py:23]: Added by 3c83e894dabde0d2f76eaef8661bb8a82d233d2b

File: python-lib/data_loader.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexcombessie) [reply python-lib/data_loader.py]: Added a comment on 60f5f7ae3ef0773c4c98be8c7cda9dcfd1a73206

**Pull request merged**